boards: qemu_arc: fix of MPU regions setup for !XIP config

kokas-a opened the pull request:
In case if Zephyr is configured with CONFIG_XIP=n it linked into SRAM. So RAM region should have EXECUTE permission.

Fixes: #38880

ruuddw commented:
From security point of view, I don't like enabling execute permissions for the full SRAM. How is this done for other QEMU targets? Can we have two SRAM regions, one for code and one for data?

Commit:
boards: qemu_arc: fix of MPU regions setup for !XIP config

Add some changes to ARC linker script. They make correct alignment
for ROMable region. Now regions borders are aligned with respect
to MPU settings.

Signed-off-by: Nikolay Agishev <agishev@synopsys.com>

kokas-a commented:
Hi colleagues,

I made changes in the follow way (divide SRAM into two parts) because of issue with linker script.
In non-XIP mode sections of "ramable_region" follows right after "romable_region" with no interval which is required to make correct alignment of "ramable_region".
On my assumption two solutions may looks more preferable:
- Add ".fill" section with appropriate size at the end of "romable_region". This will set "romable_region" border to correct alignment.
- Add "AT(address)" modifier to each section in "ramable_region".
Both of that solutions requires deep modification of the linker script which I tried to avoid.

Regards, Nikolay

evgeniy-paltsev commented on include/zephyr/arch/arc/v2/linker.ld:
```diff
@@ -58,7 +60,12 @@ MEMORY {
+#if defined(CONFIG_XIP)
 	SRAM  (rwx) : ORIGIN = SRAM_START,  LENGTH = SRAM_SIZE
+#else
+	SRAM_RX  (rx) : ORIGIN = SRAM_START,  LENGTH = (SRAM_SIZE / 2)
+	SRAM_RW  (rw) : ORIGIN = SRAM_START + (SRAM_SIZE / 2),  LENGTH = (SRAM_SIZE / 2)
```
Hmm, I'm wondering if the `LENGTH` is actually used somehow? 

The issue is that the ratio between `SRAM_RX` and `SRAM_RW` may differ an most likely it won't be 1:1.
Is it possible to avoid defining `LENGTH`?

evgeniy-paltsev replied:
If it's possible to not define `LENGTH` we can just do an assert of combined SRAM_RX and SRAM_RW sizes against SRAM size in the end.

kokas-a replied:
As far as I understand "LENGTH" is default keyword in description of memory blocks in scripts. And it should be defined for correct script working

ruuddw replied:
I believe the issue is not about "LENGTH", it is about the fixed half-half split. Since from the build process it should be known what the text and data sizes are, can't you avoid introducing new fixed-size linker segments?
Introduce some symbols that are correctly aligned (if not available already) would be enough, and use these to program the MPU.

ruuddw replied:
Thanks, current approach makes more sense (though needs some more work/though I believe).

kokas-a commented on include/zephyr/arch/arc/v2/mpu/arc_mpu.h:
```diff
@@ -32,6 +32,9 @@
 #define REGION_KERNEL_RAM_ATTR \
 			(AUX_MPU_ATTR_KW | AUX_MPU_ATTR_KR)
 
+#define REGION_KERNEL_RAM_EXEC_ATTR \
+			(AUX_MPU_ATTR_KW | AUX_MPU_ATTR_KR | AUX_MPU_ATTR_KE)
+
```
Agree

kokas-a commented on include/zephyr/arch/arc/v2/linker.ld:
```diff
@@ -121,12 +128,18 @@ SECTIONS {
 
 	GROUP_END(ROMABLE_REGION)
 
+/***************************************************************************
+ *
+ *
+ ****************************************************************************/
+
```
Will fix it

ruuddw commented on boards/arc/qemu_arc/arc_mpu_regions.c:
```diff
@@ -23,12 +23,35 @@ static struct arc_mpu_region mpu_regions[] = {
+#else
+			(uintptr_t)__kernel_ram_start,
 #endif
+			(uintptr_t)__kernel_ram_size,
```
Nothing conditional on USERSPACE required for kernel_ram_size? If the start is different, I don't think the kernel_ram_size indicates the end in that case.

kokas-a replied:
Yes, indeed, _image_ram_start and _image_ram_end are more suitable for measure kernel ram size. But _image_ram_start variable seems to be redefined in bss section. Probably new pair of variables should be implemented.

ruuddw commented on include/zephyr/arch/arc/v2/linker.ld:
```diff
@@ -93,6 +93,11 @@ SECTIONS {
 
 #include <zephyr/linker/common-rom.ld>
 #include <zephyr/linker/thread-local-storage.ld>
+#ifdef __MWDT_LINKER_CMD__
```
Is this change related to the !XIP issue, or a more general fix/clean-up?
There may me other optional, included snippets that are not at the correct place currently. Maybe handle in a separate issue/PR? Or, try and ensure all issues are fixed in this PR, preferably in separate commits.

kokas-a replied:
This change related to !XIP issue. Here I just place RODATA_SECTION at the end of ROMABLE region. The gist is in the following. RODATA_SECTION defined explicitly in the script and it's quite convenient to set the section  border (and whole ROMABLE region border) with MPU_ALIGNMENT macro.

evgeniy-paltsev commented on include/zephyr/arch/arc/arch.h:
```diff
@@ -344,6 +344,10 @@ static ALWAYS_INLINE void arch_nop(void)
 	__builtin_arc_nop();
 }
 
+#ifdef CONFIG_XIP
+extern char __arc_rw_sram_size[];
+#endif /* CONFIG_XIP */
```
We have this definition if CONFIG_XIP is enabled but we use and set it if XIP is disabled.

Looks like a bug.

kokas-a replied:
Fixed

evgeniy-paltsev commented on include/zephyr/arch/arc/v2/linker.ld:
```diff
@@ -112,23 +116,24 @@ SECTIONS {
 	__rodata_region_end = .;
-	MPU_ALIGN(__rodata_region_end - __rom_region_start);
+
+	 MPU_ALIGN(__rodata_region_end - __rom_region_start);
```
Alignment issue - let's drop this extra space.

If the newline isn't intended we can drop it either.

evgeniy-paltsev commented on boards/arc/qemu_arc/arc_mpu_regions.c:
```diff
@@ -23,12 +24,31 @@ static struct arc_mpu_region mpu_regions[] = {
 #endif /* CONFIG_COVERAGE_GCOV && CONFIG_USERSPACE */
 
 #if DT_REG_SIZE(DT_CHOSEN(zephyr_sram)) > 0
+
+/*
+ * In case if Zephyr is configured with CONFIG_XIP=n it linked into
+ * SRAM. So RAM region should have EXECUTE permission.
+ */
```
Le't move comment inside proper ifdef branch.

evgeniy-paltsev reviewed (commented):
Hi @kokas-a,

LGTM, few minor polishing requested.

I've tested this on qemu_arc_hs with XIP=n - all tests passing.
